File: gan2yed/gan2yed.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Konvertiert eine GanttProject-Datei (.gan) in eine yEd-GraphML-Datei (.graphml).

.gan ist XML: Tasks unter <tasks><task id="..." name="...">, Abhängigkeiten
unter <depend id="predecessor_id" .../> (Vorgänger -> aktueller Task = Kante).

Verwendung:
  python gan2yed.py <eingabe.gan> [ausgabe.graphml]
  Ohne ausgabe.graphml wird die Ausgabe neben der Eingabe mit Suffix .graphml geschrieben.
"""

import sys
import xml.etree.ElementTree as ET
from pathlib import Path

# GraphML-/yEd-Namespaces
GRAPHML_NS = "http://graphml.graphdrawing.org/xmlns"
YED_NS = "http://www.yworks.com/xml/schema/graphml/1.1/yed"
NS = {"g": GRAPHML_NS, "y": YED_NS}


def _register_namespaces():
    ET.register_namespace("", GRAPHML_NS)
    ET.register_namespace("y", YED_NS)


def _find_tasks_and_edges(project_root):
    """Sammelt alle Tasks (id, name) und Kanten (predecessor_id -> task_id)."""
    tasks = {}   # id -> name
    edges = []   # (source_id, target_id)

    def walk(parent_elem, parent_path=""):
        for task in parent_elem.findall("task"):
            tid = task.get("id")
            name = task.get("name") or ""
            if tid is not None:
                tasks[tid] = (name.strip() or f"Task {tid}")
            for depend in task.findall("depend"):
                dep_id = depend.get("id")
                if dep_id is not None and tid is not None:
                    edges.append((dep_id, tid))
            walk(task, parent_path + "/" + str(tid) if parent_path else str(tid))

    tasks_elem = project_root.find("tasks")
    if tasks_elem is not None:
        walk(tasks_elem)

    return tasks, edges


def _escape(s):
    if s is None:
        return ""
    return (
        str(s)
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
    )


def write_graphml(tasks, edges, out_path):
    """Schreibt eine yEd-kompatible GraphML-Datei."""
    _register_namespaces()
    root = ET.Element(
        "{" + GRAPHML_NS + "}graphml",
        attrib={
            "xmlns": GRAPHML_NS,
            "xmlns:y": YED_NS,
        },
    )
    # Key für Knotengrafik (yEd: Label + Form)
    key_nodegraphics = ET.SubElement(
        root,
        "{" + GRAPHML_NS + "}key",
        attrib={
            "id": "d0",
            "for": "node",
            "yfiles.type": "nodegraphics",
        },
    )
    graph = ET.SubElement(
        root,
        "{" + GRAPHML_NS + "}graph",
        attrib={"id": "G", "edgedefault": "directed"},
    )

    for nid, name in tasks.items():
        node = ET.SubElement(graph, "{" + GRAPHML_NS + "}node", attrib={"id": str(nid)})
        data = ET.SubElement(node, "{" + GRAPHML_NS + "}data", attrib={"key": "d0"})
        shape = ET.SubElement(data, "{" + YED_NS + "}ShapeNode")
        nlabel = ET.SubElement(shape, "{" + YED_NS + "}NodeLabel")
        nlabel.text = name

    for i, (src, tgt) in enumerate(edges):
        if src in tasks and tgt in tasks:
            ET.SubElement(
                graph,
                "{" + GRAPHML_NS + "}edge",
                attrib={"id": f"e{i}", "source": str(src), "target": str(tgt)},
            )

    tree = ET.ElementTree(root)
    ET.indent(tree, space="  ")
    with open(out_path, "wb") as f:
        tree.write(
            f,
            encoding="utf-8",
            xml_declaration=True,
            default_namespace=GRAPHML_NS,
            method="xml",
        )


def main():
    if len(sys.argv) < 2:
        print(__doc__.strip().split("\n\n")[0], file=sys.stderr)
        print("Verwendung: python gan2yed.py <eingabe.gan> [ausgabe.graphml]", file=sys.stderr)
        sys.exit(1)

    in_path = Path(sys.argv[1])
    if not in_path.exists():
        print(f"Datei nicht gefunden: {in_path}", file=sys.stderr)
        sys.exit(2)

    out_path = Path(sys.argv[2]) if len(sys.argv) > 2 else in_path.with_suffix(".graphml")

    try:
        tree = ET.parse(in_path)
        root = tree.getroot()
    except ET.ParseError as e:
        print(f"XML-Fehler in {in_path}: {e}", file=sys.stderr)
        sys.exit(3)

    tasks, edges = _find_tasks_and_edges(root)
    if not tasks:
        print("Keine Tasks in der GanttProject-Datei gefunden.", file=sys.stderr)
        sys.exit(4)

    write_graphml(tasks, edges, out_path)
    print(f"Geschrieben: {out_path} ({len(tasks)} Knoten, {len(edges)} Kanten)")


if __name__ == "__main__":
    main()

File: gan2yed/gan2yed.rb
#!/usr/bin/env ruby
# encoding: utf-8
#
# Konvertiert eine GanttProject-Datei (.gan) in eine yEd-GraphML-Datei (.graphml).
#
# .gan ist XML: Tasks unter <tasks><task id="..." name="...">, Abhängigkeiten
# unter <depend id="predecessor_id" .../> (Vorgänger -> aktueller Task = Kante).
#
# Verwendung:
#   ruby gan2yed.rb <eingabe.gan> [ausgabe.graphml]
#   Ohne ausgabe.graphml wird die Ausgabe neben der Eingabe mit Suffix .graphml geschrieben.
#

require 'rexml/document'
require 'rexml/formatters/pretty'

GRAPHML_NS = 'http://graphml.graphdrawing.org/xmlns'
YED_NS     = 'http://www.yworks.com/xml/schema/graphml/1.1/yed'

def find_tasks_and_edges(doc)
  tasks = {}
  edges = []
  doc.elements.each('//task') do |task|
    tid = task.attributes['id']
    next unless tid
    name = (task.attributes['name'] || '').strip
    tasks[tid] = name.empty? ? "Task #{tid}" : name
    task.elements.each('depend') do |dep|
      dep_id = dep.attributes['id']
      edges << [dep_id, tid] if dep_id
    end
  end
  [tasks, edges]
end

def write_graphml(tasks, edges, out_path)
  doc = REXML::Document.new
  doc << REXML::XMLDecl.new('1.0', 'UTF-8')

  root = REXML::Element.new('graphml')
  root.add_attribute('xmlns', GRAPHML_NS)
  root.add_attribute('xmlns:y', YED_NS)
  doc.add_element(root)

  key = REXML::Element.new('key')
  key.add_namespace(GRAPHML_NS)
  key.add_attribute('id', 'd0')
  key.add_attribute('for', 'node')
  key.add_attribute('yfiles.type', 'nodegraphics')
  root.add_element(key)

  graph = REXML::Element.new('graph')
  graph.add_namespace(GRAPHML_NS)
  graph.add_attribute('id', 'G')
  graph.add_attribute('edgedefault', 'directed')
  root.add_element(graph)

  tasks.each do |nid, name|
    node = REXML::Element.new('node')
    node.add_namespace(GRAPHML_NS)
    node.add_attribute('id', nid.to_s)
    graph.add_element(node)

    data = REXML::Element.new('data')
    data.add_namespace(GRAPHML_NS)
    data.add_attribute('key', 'd0')
    node.add_element(data)

    shape = REXML::Element.new('ShapeNode')
    shape.add_namespace('y', YED_NS)
    data.add_element(shape)

    nlabel = REXML::Element.new('NodeLabel')
    nlabel.add_namespace('y', YED_NS)
    nlabel.text = name
    shape.add_element(nlabel)
  end

  edges.each_with_index do |(src, tgt), i|
    next unless tasks[src] && tasks[tgt]
    edge = REXML::Element.new('edge')
    edge.add_namespace(GRAPHML_NS)
    edge.add_attribute('id', "e#{i}")
    edge.add_attribute('source', src.to_s)
    edge.add_attribute('target', tgt.to_s)
    graph.add_element(edge)
  end

  formatter = REXML::Formatters::Pretty.new(2)
  formatter.compact = true
  File.open(out_path, 'w:UTF-8') { |f| formatter.write(doc, f) }
end

def main
  if ARGV.empty?
    warn 'Konvertiert eine GanttProject-Datei (.gan) in eine yEd-GraphML-Datei (.graphml).'
    warn 'Verwendung: ruby gan2yed.rb <eingabe.gan> [ausgabe.graphml]'
    exit 1
  end

  in_path = ARGV[0]
  unless File.file?(in_path)
    warn "Datei nicht gefunden: #{in_path}"
    exit 2
  end

  out_path = ARGV[1] || in_path.sub(/\.gan\z/i, '.graphml')

  begin
    doc = REXML::Document.new(File.read(in_path))
  rescue REXML::ParseException => e
    warn "XML-Fehler in #{in_path}: #{e.message}"
    exit 3
  end

  tasks, edges = find_tasks_and_edges(doc)
  if tasks.empty?
    warn 'Keine Tasks in der GanttProject-Datei gefunden.'
    exit 4
  end

  write_graphml(tasks, edges, out_path)
  puts "Geschrieben: #{out_path} (#{tasks.size} Knoten, #{edges.size} Kanten)"
end

main if __FILE__ == $PROGRAM_NAME
